ext/backtracie_native_extension/backtracie.c

```diff
@@ -16,105 +16,116 @@
-static VALUE collect_backtrace_locations(VALUE self, VALUE thread, int ignored_stack_top_frames) {
-  int stack_depth = 0;
-  raw_location raw_locations[MAX_STACK_DEPTH];
-
-  if (thread == Qnil) {
-    // Get for own thread
-    stack_depth = backtracie_rb_profile_frames(MAX_STACK_DEPTH, raw_locations);
-  } else {
-    stack_depth = backtracie_rb_profile_frames_for_thread(thread, MAX_STACK_DEPTH, raw_locations);
-    if (stack_depth == 0 && !backtracie_is_thread_alive(thread)) return Qnil;
+// Get array of Backtracie::Locations for a given thread; if thread is nil,
+// returns for the current thread
+static VALUE collect_backtrace_locations(VALUE self, VALUE thread,
+                                         int ignored_stack_top_frames) {
+  if (!RTEST(thread)) {
+    thread = rb_thread_current();
+  }
+  int raw_frame_count = backtracie_frame_count_for_thread(thread);
+
+  // Allocate memory for the raw_locations, and keep track of it on the Ruby
+  // heap so it will be GC'd even if we raise.
+  // Zero the frame array so our mark function doesn't get confused too.
+  raw_location *raw_frames = calloc(raw_frame_count, sizeof(raw_location));
+  VALUE locs_gc = Data_Wrap_Struct(0, backtracie_frame_mark, free, raw_frames);
+  VALUE rb_locations = rb_ary_new_capa(raw_frame_count);
+
+  for (int i = ignored_stack_top_frames; i < raw_frame_count; i++) {
+    raw_frames[i].abi_version = BACKTRACIE_ABI_VERSION;
+    backtracie_capture_frame_for_thread(thread, i, &raw_frames[i]);
   }
 
-  VALUE locations = rb_ary_new_capa(stack_depth - ignored_stack_top_frames);
-
-  // MRI does not give us the path or line number for frames implemented using C code. The convention in
-  // Kernel#caller_locations is to instead use the path and line number of the last Ruby frame seen.
-  // Thus, we keep that frame here to able to replicate that behavior.
-  // (This is why we also iterate the frames array backwards below -- so that it's easier to keep the last_ruby_frame)
-  raw_location *last_ruby_location = 0;
-
-  for (int i = stack_depth - 1; i >= ignored_stack_top_frames; i--) {
-    VALUE location = Qnil;
-
-    if (raw_locations[i].is_ruby_frame) {
-      last_ruby_location = &raw_locations[i];
-      location = ruby_frame_to_location(&raw_locations[i]);
-    } else {
-      location = cfunc_frame_to_location(&raw_locations[i], last_ruby_location);
+  for (int i = ignored_stack_top_frames; i < raw_frame_count; i++) {
+    if (raw_frames[i].is_valid) {
+      VALUE rb_loc = frame_to_location(raw_frames, raw_frame_count, i);
+      rb_ary_push(rb_locations, rb_loc);
     }
+  }
 
-    rb_ary_store(locations, i - ignored_stack_top_frames, location);
+  // To maintain compatability with the Ruby thread backtrace behavior, if a
+  // thread's stack is empty (because its dead) then return nil, not empty.
+  if (rb_array_len(rb_locations) == 0) {
+    rb_locations = Qnil;
   }
```

Comment:
ah OK, i'll put the explicit check for thread liveness back in then.

Reply:
For dd-trace-rb I actually detect the thread-in-native situation and even add a single placeholder frame, as some pprof viewers don't really like the whole "an empty stack has counts assigned" -- https://github.com/DataDog/dd-trace-rb/blob/f711887c2861facde72b86bf7468bc5be0b0ade7/ext/ddtrace_profiling_native_extension/collectors_stack.c#L209-L229

Reply:
Hmm, interesting! I _think_ adding the placeholder frame would be the responsibility of a profiler using backtracie, rather than backtracie itself, right? I'll go do a similar thing in ruby_memprofiler_pprof :)